openwrt-openwisp-monitoring/files/monitoring.agent

```diff
@@ -78,24 +65,140 @@ collect_data(){
-  if [ "$response_code" = "200" ]; then
-    [ "$verbose_mode" = "1" ] && logger -s "Data sent successfully." \
-                                        -p daemon.info
-  else
-    logger -s "Data not sent successfully. Response code is $response_code" \
+set_url_and_curl(){
+  if [ -z "$BASE_URL" ]; then
+    logger -s "missing required --url option" \
            -t openwisp_monitoring \
            -p daemon.err
+    exit 1
+  fi
 
-    [ "$verbose_mode" = "0" ] && logger -s "Run with verbose mode to find more." \
-                                        -t openwisp_monitoring \
-                                        -p daemon.err
+  if [ -z "$UUID" ]; then
+    logger -s "missing required --uuid option" \
+            -t openwisp_monitoring \
+            -p daemon.err
+    exit 1
   fi
-  sleep "$INTERVAL" & wait $!
-done
+
+  if [ -z "$KEY" ]; then
+    logger -s "missing required --key option" \
+           -t openwisp_monitoring \
+           -p daemon.err
+    exit 1
+  fi
+
+  URL="$BASE_URL/api/v1/monitoring/device/$UUID/?key=$KEY"
+
+  CURL_COMMAND="curl -s -w "%{http_code}""
+  [ "$VERIFY_SSL" = "0" ] && CURL_COMMAND="$CURL_COMMAND -k"
+  [ "$verbose_mode" = "1" ] && CURL_COMMAND="$CURL_COMMAND -v"
+  echo "0"
+}
+
+save_data() {
+  # Remove double quotes from interfaces
+  MONITORED_INTERFACES=$(echo "$MONITORED_INTERFACES" | tr -d '"')
+
+  while true
+  do
+    data="$(collect_data)"
+    memory_available="$(check_available_memory)"
+    if [ "$memory_available" != "0" ]; then
+        file_name="$(date -u +'%d-%m-%Y_%H:%M:%S').000000.json" # extra zeroes are added for nanoseconds precision
+        # make directory
+        mkdir -p "$TMP_DIR"
+        #save data with file_name
+        echo "$data" > "$TMP_DIR/$file_name"
+        [ "$verbose_mode" = "1" ] && logger -s "Data saved temporarily" \
+                                            -p daemon.info
+    fi
+    # get process id of container sending data
+    pid=$(pgrep -f "openwisp_monitoring.*uuid")
+    kill -SIGUSR1 "$pid"
+    sleep "$INTERVAL"
+  done
+}
+
+handle_sigusr1() {
+  if [ "$verbose_mode" = "0" ]; then
+    logger -s "SIGUSR1 received! Sending data" \
+           -t openwisp_monitoring \
+           -p daemon.info
+  else
+    logger -s "SIGUSR1 received! Sending data" \
+           -p daemon.info
+  fi
+}
+
+send_data() {
+  while true
+  do
+    for file in "$TMP_DIR"/*.json
+    do
+      if [ ! -f "$file" ]; then
+        [ "$verbose_mode" = "1" ] && logger -s "No data file found to send." \
+                                            -p daemon.info
+        trap handle_sigusr1 USR1
+        # SIGUSR1 signal received interrupt sleep and continue sending data
+        sleep "$INTERVAL" & wait $!
+        trap "" USR1
+        continue
+      fi
+      basefilename=${file##*/}
+      filename=${basefilename%.*}
+      url="$URL&time=$filename"
+      data=$(cat "$file")
+      # retry sending data in case of failure
+      MAX_RETRY=${MAX_RETRY:-5}
```

Comment:
2 observations here:

- isn't this value a constant which doesn't change in the loop?
  if so, why define it inside loop, which is executed many times? Can't this be defined just once at the beginning of the function?
- I think it can be renamed to `MAX_RETRIES`

Keep in mind the same suggestion for any other variable which doesn't need to be in the loop, I see a few different ones here so please double check here and in any other loop.

Please also keep in mind that uppercase is generally used for constants which are defined at the beginning of the program and used in different places. For anything else defined only in specific functions, please use lowercase naming and flag them as [local variables](https://tldp.org/LDP/abs/html/localvar.html).